src/nn/add_into.rs

```diff
@@ -0,0 +1,238 @@
+    }
+}
+
+impl<T: ResetParams> ResetParams for AddInto<T> {
+    fn reset_params<R: rand::Rng>(&mut self, rng: &mut R) {
+        self.0.reset_params(rng);
+    }
+}
+
+macro_rules! tuple_impls {
+    ($head:ident $headin:ident [$($tails:ident $tailsin:ident),+]) => {
+        impl<
+            Output: Tensor<Dtype = f32>,
+            $headin: Tensor<Dtype = f32>,
+            $($tailsin: Tensor<Dtype = f32>,)+
+            $head: Module<$headin, Output = Output>,
+            $($tails: Module<$tailsin, Output = Output>,)+
+        > Module<($headin, $($tailsin,)+)> for AddInto<($head, $($tails,)+)> {
+            type Output = Output;
+
+            #[allow(non_snake_case)]
+            fn forward(&self, x: ($headin, $($tailsin,)+)) -> Self::Output {
+
+                // inputs
+                let ($head, $($tails),+) = x;
+
+                // layers
+                let ($headin, $($tailsin),+) = &self.0;
+
+                // forward
+                let ($head, $($tails),+) = ($headin.forward($head), $($tailsin.forward($tails)),+);
+
+                // add together
+                $(
+                    let $head = add($head, $tails);
+                )+
+
+                $head
+            }
+        }
+
+
+        impl<
+            Output: Tensor<Dtype = f32>,
+            $headin: Tensor<Dtype = f32>,
+            $($tailsin: Tensor<Dtype = f32>,)+
+            $head: Module<$headin, Output = Output>,
+            $($tails: Module<$tailsin, Output = Output>,)+
+        > ModuleMut<($headin, $($tailsin,)+)> for AddInto<($head, $($tails,)+)> {
+            type Output = Output;
+
+            #[allow(non_snake_case)]
+            fn forward_mut(&mut self, x: ($headin, $($tailsin,)+)) -> Self::Output {
+
+                // inputs
+                let ($head, $($tails),+) = x;
+
+                // layers
+                let ($headin, $($tailsin),+) = &self.0;
+
+                // forward
+                let ($head, $($tails),+) = ($headin.forward($head), $($tailsin.forward($tails)),+);
+
+                // add together
+                $(
+                    let $head = add($head, $tails);
+                )+
+
+                $head
+            }
+        }    }
+}
+
+tuple_impls!(A Ai [B Bi]);
+tuple_impls!(A Ai [B Bi, C Ci]);
+tuple_impls!(A Ai [B Bi, C Ci, D Di]);
+tuple_impls!(A Ai [B Bi, C Ci, D Di, E Ei]);
+tuple_impls!(A Ai [B Bi, C Ci, D Di, E Ei, F Fi]);
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{nn::tests::SimpleGradients, unique_id::HasUniqueId};
+
+    #[test]
+    fn test_add_into_2() {
+        type Model = AddInto<(Linear<2, 5>, Linear<3, 5>)>;
+        let m: Model = Default::default();
+        let _: Tensor1D<5, OwnedTape> =
+            m.forward((Tensor1D::zeros().traced(), Tensor1D::zeros().traced()));
+        let _: Tensor2D<3, 5, OwnedTape> = m.forward((
+            Tensor2D::<3, 2>::zeros().traced(),
+            Tensor2D::<3, 3>::zeros().traced(),
+        ));
```

Comment:
Side comment, but I think with the way this is set up you actually only need the tape on the first input if I'm understanding the forward correctly.

Reply:
Yeah, wasn't really sure about how to handle this here, as I'm not super familliar with how tapes work.

Reply:
basically the forward expands to
```rust
// tuple, so we get all layers
let (headin, ain, bin, cin) = &self.0;

// forward
let (head, a, b, c) = (headin.forward(), ain.forward(), bin.forward(), cin.forward());

// add all of them together recursively
let head = add(head, a);
let head = add(head, b);
let head = add(head, c);

head
```

Reply:
hmm, looking at the add and binary_map functions it seems those merge the tapes, so maybe it's not needed? 

Both this and doing the with_new_tape on the input tails before doing forward passes the test_missing_gradients test

Reply:
Ahh yeah the expansion is helpful, so the ain/bin/cin forwards aren't recorded unless their inputs have tapes. Should probably think about whether this should require that all inputs have tapes or not - ideally this could work in both cases. That can probably be done in a separate follow up PR though, I think expecting all of them to have tapes at first is reasonable enough

Reply:
Yeah makes sense. 
So should I change it to have it expect tapes for all the inputs?

Reply:
Nah I think it's good how it is, I'll just make an issue based on this thread for tracking purposes. I think this is good to merge though!